File: tests_end_to_end/page_objects/BasePage.py
from playwright.sync_api import Page
import os


class BasePage:
    def __init__(self, page: Page, path: str, query_params: str = ""):
        """
        Base page class that handles workspace configuration for all pages.

        Args:
            page: Playwright page object
            path: The path part of the URL (e.g., 'projects', 'traces', etc.)
            query_params: Optional query parameters to append to the URL (e.g., '?tab=feedback-definitions')
        """
        self.page = page
        self.workspace = os.environ.get("OPIK_WORKSPACE", "default")
        self.base_url = os.environ.get("OPIK_BASE_URL", "http://localhost:5173")

        # Remove leading/trailing slashes and combine path components
        clean_path = path.strip("/")
        self.path = f"{self.workspace}/{clean_path}"
        if query_params:
            # Ensure query params start with '?' if provided
            if not query_params.startswith("?"):
                query_params = f"?{query_params}"
            self.path = f"{self.path}{query_params}"

    def go_to_page(self):
        """Navigate to the page URL"""
        # Combine base URL with path, ensuring no double slashes
        full_url = f"{self.base_url.rstrip('/')}/{self.path}"
        self.page.goto(full_url)

File: tests_end_to_end/tests_end_to_end_ts/typescript-tests/page-objects/base.page.ts
/**
 * Base Page Object
 *
 * Mirrors the Python BasePage class from page_objects/BasePage.py
 * Handles workspace-aware navigation for all pages
 */

import { Page } from '@playwright/test';
import { getEnvironmentConfig } from '../config/env.config';

export class BasePage {
  protected page: Page;
  protected workspace: string;
  protected baseUrl: string;
  protected path: string;

  constructor(page: Page, path: string, queryParams: string = '') {
    this.page = page;

    const envConfig = getEnvironmentConfig();
    const envData = envConfig.getConfig();

    this.workspace = envData.workspace;
    this.baseUrl = envData.baseUrl;

    // Remove leading/trailing slashes and combine path components
    const cleanPath = path.replace(/^\/+|\/+$/g, '');
    this.path = `${this.workspace}/${cleanPath}`;

    // Add query params if provided
    if (queryParams) {
      if (!queryParams.startsWith('?')) {
        queryParams = `?${queryParams}`;
      }
      this.path = `${this.path}${queryParams}`;
    }
  }

  /**
   * Navigate to the page URL
   */
  async goto(): Promise<void> {
    const fullUrl = `${this.baseUrl.replace(/\/+$/, '')}/${this.path}`;
    await this.page.goto(fullUrl);
  }

  /**
   * Get the full URL for this page
   */
  getUrl(): string {
    return `${this.baseUrl.replace(/\/+$/, '')}/${this.path}`;
  }
}
